pkg/telemetry/platform.go

```diff
@@ -0,0 +1,99 @@
+package telemetry
+
+import (
+	"strings"
+
+	v1 "k8s.io/api/core/v1"
+)
+
+const (
+	openshiftIdentifier = "node.openshift.io/os_id"
+	k3sIdentifier       = "k3s"
+	awsIdentifier       = "aws"
+	gkeIdentifier       = "gce"
+	azureIdentifier     = "azure"
+	kindIdentifier      = "kind"
+	rancherIdentifier   = "cattle-system"
+)
+
+func CollectK8sPlatform(node v1.Node, namespaces v1.NamespaceList) string {
+	if result := isMultiplePlatforms(node, namespaces); result != "" {
+		return result
+	}
+
+	if isAWSPlatform(node) {
```

Comment:
Another way to improve this part is something like below. In this case, for every new platform, you will only need to add new function to the slice:
```go
type k8sState struct {
	node       v1.Node
	namespaces v1.NamespaceList
}

type platformExtractor func(k8sState) (ClusterPlatform, bool)

var platformExtractors = []platformExtractor{
	func(state k8sState) (ClusterPlatform, bool) {
		if strings.HasPrefix(state.node.Spec.ProviderID, "gce") {
			return ClusterPlatformGKE, true
		}
		return "", false
	},
	func(state k8sState) (ClusterPlatform, bool) {
		if strings.HasPrefix(state.node.Spec.ProviderID, "aws") {
			return ClusterPlatformEKS, true
		}
		return "", false
	},
}

func GetPlatform(node v1.Node, namespaces v1.NamespaceList) ClusterPlatform {
	state := k8sState{
		node:       node,
		namespaces: namespaces,
	}
	
	for _, extractor := range platformExtractors {
		if platform, ok := extractor(state); ok {
			return platform
		}
	}

	return ClusterPlatformUnknown
}
```

Reply:
@pleshakov I see this version as doing the same thing as the other version you recommended with the map, but with more complexity. For my understanding, what are the benefits that we would gain by doing this versus doing the suggestion with a map?

Reply:
@bjee19 
this version is easier to extend - to add any new platforms later.

More over, let's not forget other platforms that don't support providerID - they can also all fit here:

```go
var platformExtractors = []platformExtractor{
        openShiftExtractor, // need to define this and similar functions separately....
        rancherExtractor,
	buildProviderIDExtractor("gce", ClusterPlatformGKE),
	buildProviderIDExtractor("eks", ClusterPlatformEKS),
	// and all other provider id ...s
	unknownProviderIDExtractor, // here - extract the id of the provider as a fallback
}

```

Reply:
Ah I see, does the real benefit come with other platforms that don't support providerID? 

Because if we're just dealing with those that support providerID, the other version would just require adding a line to the `mapProviderIDToPlatform` so in my opinion both would be equally easy to extend, but this one would just be more complex.

Reply:
> Ah I see, does the real benefit come with other platforms that don't support providerID?

yeah, so all those extractors are called similarly, so that we can add more or remove

this way you will not have a function with many ifs like in your original code